samples: pmic: npmx: add buck sample (#11158)
nrfconnect/sdk-nrf: samples/pmic/npmx/buck/README.rst

.. _nPM1300 BUCK:

> nPM1300 BUCK
> ############

Reviewer: ```suggestion
nPM1300: BUCK
#############
```

Overview

> ********

Reviewer 2: Sample is not matching the template: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/templates/sample_README.html

> This sample demonstrates the usage of nPM1300 PMIC BUCKs.
> For testing connect voltage multimeter to VOUT1 (output voltage pin of BUCK1). In this example
> voltage can be changed only for BUCK1. BUCK2 voltage is set by default to 3.0V.
> In main.c file select your test case to be executed.
> You can also modify expected output voltages and selected GPIOs to control the system.

Reviewer: This sentence should be the very first sentence in this file (line 6):
```suggestion
The BUCK sample demonstrates how to ... using nPM1300 PMIC.
```

Next sentences should be the first sentences of **Building and running** or **Testing** section - please decide where current testcases' descriptions should be placed and what is missing. Once done, I'll convert them into steps.
```suggestion

For testing purposes, do the following:

#. Connect a voltage multimeter to **VOUT1** that is the output voltage pin of BUCK1 mode.

.. note::
   In this sample, voltage can be changed only for BUCK1 mode.
   BUCK2 voltage is set by default to 3.0V.
   
#. Go to the main.c file and select your test case to be executed.

You can also modify expected output voltages and selected GPIOs to control the system.
```

> Testcases
> ---------

Reviewer: ```suggestion
Building and running
**********************
```

> Testcases
> ---------

Reviewer: ```suggestion
Building and running
**********************
```

Reviewer: Based on our [template](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/templates/sample_README.html), please decide what should be called **Testing** indeed, and what should be placed in **Building and running** part.

Test set output BUCK voltage
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> 

Reviewer: Each testcase starts with "In testcase selection..." - what should I do after wiring to get to this point?

Test retention voltage using GPIO
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In `Testcase selection` select `Set retention voltage of buck 1 with selected external pin`.
Now only test_retention_voltage() function is called.
GPIO1 (`NPMX_GPIO_INSTANCE_1`) is set to input mode and configured to be used as a retention input selector for buck 1 (`NPMX_BUCK_INSTANCE_1`).
Normal mode voltage is set to 1.7 V (`NPMX_BUCK_VOLTAGE_1V7`) and retention voltage to 3.3 V (`NPMX_BUCK_VOLTAGE_3V3`).
When you change the input state of the selected GPIO you should see voltage changes in the buck 1 output.

Test enable BUCK using GPIO
~~~~~~~~~~~~~~~~~~~~~~~~~~~

In `Testcase selection` select `Set enable mode of buck 1 with selected external pin`.
Now only test_enable_bucks() function is called.
GPIO3 (`NPMX_GPIO_INSTANCE_3`) is set input mode and configured to be used as enable mode selector for buck 1 (`NPMX_BUCK_INSTANCE_1`).
The output voltage of buck 1 is set to 3.3 V (`NPMX_BUCK_VOLTAGE_3V3`).
When you change the input state of the selected GPIO you should see voltage changes in the buck 1 output - from 0 V to previously selected voltage.

Requirements
************

> To run this sample, do the following:

Reviewer: If these are the steps that should be done to get to the testcases, they should be listed as the first steps in the **Building and running** section.

* Connect the nRF5340 or nRF52840 DK to a host device.
* Open a COM port.
* Connect nPM1300 EK with the chosen DK.

> Connections
> -----------

Reviewer: ```suggestion
Wiring
******
```

This entire section should go under the **Overview** one (as indicated in the template).

.. list-table:: nPM1300 EK connections.
   :widths: 25 25 25
   :header-rows: 1

   * - nPM1300 EK
     - nRF5340 DK
     - nRF52840 DK
   * - GPIO0
     - P1.10
     - P1.10
   * - SDA
     - P1.02
     - P0.26
   * - SCL
     - P1.03
     - P0.27
   * - VOUT2 + GND
     - External supply (P21)
     - External supply (P21)

> On nPM1300 EK do the following:
> 
> * Connect USB to power supply (J3).
> * On **P18** connect with the VOUT2 - VDDIO jumper.
> * On **P2** connect with the VBAT - VBATIN jumper.
> * On **P17** connect all LEDs with jumpers.
> * On **P13** connect with the RSET1 - GND jumper.
> * On **P14** connect with the RSET2 - VSET2 jumper.

Reviewer: Copy-pasted from the Fuel gauge sample:

```suggestion
Make the following connections on the nPM1300 EK:

* Connect a USB power supply to the **J3** connector.
* Connect a suitable battery to the **J2** connector.
* On the **P18** pin header, connect **VOUT2** and **VDDIO** pins with a jumper.
* On the **P2** pin header, connect **VBAT** and **VBATIN** pins with a jumper.
* On the **P17** pin header, connect all LEDs with jumpers.
* On the **P13** pin header, connect **RSET1** and **VSET1** pins with a jumper.
* On the **P14** pin header, connect **RSET2** and **VSET2** pins with a jumper.
```

> On nRF5340 DK do the following:

Reviewer: As we have two options - one of two DKs to be chosen - I suggest creating tabs for them, similar to that:

![image](https://github.com/nrfconnect/sdk-nrf/assets/115017424/449997dc-cb18-4273-a590-e198996b03af)

If you agree, I'll do it.

Steps should be a little bit more descriptive - as above of EK.

  * set BUCK output voltage
  * set BUCK retention voltage with GPIO
  * control BUCK enable mode with GPIO

Wiring
******

Connect the TWI interface between the chosen DK and the nPM1300 EK as in the following table:

.. list-table:: nPM1300 EK connections.
   :widths: 25 25 25
   :header-rows: 1

   * - nPM1300 EK
     - nRF5340 DK
     - nRF52840 DK
   * - GPIO0
     - P1.10
     - P1.10
   * - SDA
     - P1.02
     - P0.26
   * - SCL
     - P1.03
     - P0.27
   * - VOUT2 & GND
     - External supply (P21)
     - External supply (P21)

Make the following connections on the nPM1300 EK:

* Connect a USB power supply to the **J3** connector.
* Connect a suitable battery to the **J2** connector.
* On the **P18** pin header, connect **VOUT2** and **VDDIO** pins with a jumper.
* On the **P2** pin header, connect **VBAT** and **VBATIN** pins with a jumper.
* On the **P17** pin header, connect all LEDs with jumpers.
* On the **P13** pin header, connect **RSET1** and **GND** pins with a jumper.
* On the **P14** pin header, connect **RSET2** and **VSET2** pins with a jumper.

Make the following connections on the DK:

* Switch nRF power source **SW9** to **VDD**.

> * Switch VEXT -> nRF **SW10** to **ON**.

Author: https://infocenter.nordicsemi.com/topic/ug_nrf52840_dk/UG/dk/images/nRF52840/pca10056_front.svg
It's bottom left, SW10. Not sure how to describe it better.

Configuration
*************

|config|

Configuration options
=====================

Check and configure the following Kconfig options to specify which testcase to run:

> .. TESTCASE_SET_BUCK_VOLTAGE:

Reviewer 3: Add CONFIG_ in front of all Kconfig options.

TESTCASE_SET_BUCK_VOLTAGE
  This sample configuration enables simple output voltage testcase.
  Voltage is set to 2.4 V (`NPMX_BUCK_VOLTAGE_2V4`) on BUCK 1.

.. TESTCASE_OUTPUT_VOLTAGE:

TESTCASE_OUTPUT_VOLTAGE

>   This sample configuration enables advanced output voltage testcase.
>   BUCK 1 is starting from voltage 1.0 V (`NPMX_BUCK_VOLTAGE_1V0`).
>   The output voltage will increase from the selected to the maximum voltage (3.3 V).

Reviewer: ```suggestion
  This sample configuration enables an advanced output voltage testcase.
  The BUCK1 mode starts from voltage 1.0 V (`NPMX_BUCK_VOLTAGE_1V0`).
  The output voltage increases from the selected voltage to the maximum one (3.3 V).
```

.. TESTCASE_RETENTION_VOLTAGE:

TESTCASE_RETENTION_VOLTAGE

>   This sample configuration enables retention voltage testcase.
>   GPIO1 (`NPMX_GPIO_INSTANCE_1`) is set to input mode and configured to be used as a retention input selector for BUCK 1 (`NPMX_BUCK_INSTANCE_1`).
>   Normal mode voltage is set to 1.7 V (`NPMX_BUCK_VOLTAGE_1V7`) and retention voltage to 3.3 V (`NPMX_BUCK_VOLTAGE_3V3`).
>   When you change the input state of the selected GPIO you should see voltage changes in the BUCK 1 output.

Reviewer: ```suggestion
  This sample configuration enables a retention voltage testcase.
  GPIO1 (`NPMX_GPIO_INSTANCE_1`) is set to the input mode and is configured to be used as a retention input selector for the BUCK1 mode (`NPMX_BUCK_INSTANCE_1`).
  The normal mode voltage is set to 1.7 V (`NPMX_BUCK_VOLTAGE_1V7`) and the retention voltage to 3.3 V (`NPMX_BUCK_VOLTAGE_3V3`).
  When you change the input state of the selected GPIO, the voltage in the BUCK1 output will change.
```

.. TESTCASE_ENABLE_BUCKS_VIA_PIN:

> TESTCASE_ENABLE_BUCKS_VIA_PIN

Reviewer: Is it possible to change these names to avoid using **via** word? Can we change it to **through**?

>   GPIO3 (`NPMX_GPIO_INSTANCE_3`) is set input mode and configured to be used as enable mode selector for BUCK 1 (`NPMX_BUCK_INSTANCE_1`).
>   The output voltage of BUCK 1 is set to 3.3 V (`NPMX_BUCK_VOLTAGE_3V3`).
>   When you change the input state of the selected GPIO you should see voltage changes in the BUCK 1 output - from 0 V to previously selected voltage.

Reviewer: ```suggestion
  GPIO3 (`NPMX_GPIO_INSTANCE_3`) is set to the input mode and is configured to be used as an enable mode selector for the BUCK1 mode (`NPMX_BUCK_INSTANCE_1`).
  The output voltage of the BUCK1 mode is set to 3.3 V (`NPMX_BUCK_VOLTAGE_3V3`).
  When you change the input state of the selected GPIO, the voltage in the BUCK1 output will change from 0 V to the previously selected voltage.
```

.. note::

  Only one Kconfig option may be enabled at a time.

Building and running
********************

.. |sample path| replace:: :file:`samples/pmic/npmx/buck`

.. include:: /includes/build_and_run.txt

Testing
=======

> |test_sample|
> 
> To verify voltage, connect multimeter to VOUT1 (output voltage pin of BUCK1).
> Voltage can be changed only for BUCK1. BUCK2 voltage is set by default to 3.0 V.
> 
> #. |connect_kit|
> #. |connect_terminal|
> #. For simple output voltage testcase: Check output voltage on EK VOUT1 - it should be 2.4 V.
> #. For advanced output voltage testcase: Check output voltage on EK VOUT1 - it should start at 1.0 V and increase every 200 ms up to 3.3 V.
> #. For retention voltage testcase: Connect EK GPIO1 to GND. Check output voltage on EK VOUT1 - it should be 1.7 V. Connect EK GPIO1 to VDD. Check output voltage on EK VOUT1 - it should be 3.3 V.
> #. For enable bucks via pin testcase: Connect EK GPIO3 to GND. Check output voltage on EK VOUT1 - it should be 0.0 V. Connect EK GPIO3 to VDD. Check output voltage on EK VOUT1 - it should be 3.3 V.

Reviewer: ```suggestion
After programming the sample to your development kit, test it by verifying the voltage.

#. Connect a multimeter to VOUT1 that is the output voltage pin of BUCK1.
#. |connect_kit|
#. |connect_terminal|
#. Depending on the chosen testcase, do the following:

   #. For the simple output voltage testcase, check the output voltage on EK VOUT1 - it should be 2.4 V.
   #. For the advanced output voltage testcase, check the output voltage on EK VOUT1 - it should start at 1.0 V and increase every 200 ms up to 3.3 V.
   #. For the retention voltage testcase, connect EK GPIO1 to GND, and check the output voltage on EK VOUT1 - it should be 1.7 V. Next, connect EK GPIO1 to VDD, and check the output voltage on EK VOUT1 - it should be 3.3 V.
   #. For the enabling bucks through a pin testcase, connect EK GPIO3 to GND, and check the output voltage on EK VOUT1 - it should be 0.0 V. Next, connect EK GPIO3 to VDD, and check the output voltage on EK VOUT1 - it should be 3.3 V.
```

As the steps above describe a way to test/check sth, and we don't change anything here, I assume that the sentence pasted below should be placed somewhere else:
Voltage can be changed only for BUCK1. BUCK2 voltage is set by default to 3.0 V.

Dependencies
************

> This sample uses drivers from npmx.

Reviewer: More that one, i.e., the BUCK one?

```suggestion
This sample uses the :file:`npmx_buck.h` npmx driver.
```

Reviewer 3: I guess the header files are not drivers.

Author: Could be added when @b-gent confirms that npmx docs are linkable.

.. _npmx_buck_sample:

npmx: BUCK
##########

.. contents::
   :local:
   :depth: 2

The npmx BUCK sample demonstrates how to control PMIC BUCKs using npmx drivers.

Requirements
************

The sample supports the following development kits:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf52840dk_nrf52840, nrf5340dk_nrf5340_cpuapp

The sample also requires an nPM1300 EK.

Overview
********

> This sample uses npmx BUCK driver to:

Reviewer: ```suggestion
This sample uses the npmx BUCK driver to do the following:
  * Set output voltage for the BUCK converter. 
  * Set retention voltage for the BUCK converter with GPIO.
  * Control enable mode for the BUCK converter with GPIO.
```

> This sample uses npmx BUCK driver to:
>   * set output voltage for BUCK converter
>   * set retention voltage for BUCK converter with GPIO
>   * control enable mode for BUCK converter with GPIO

Reviewer: ```suggestion
This sample uses the npmx BUCK driver to do the following:
  * Set output voltage for the BUCK converter. 
  * Set retention voltage for the BUCK converter with GPIO.
  * Control enable mode for the BUCK converter with GPIO.
```

>   * control enable mode for BUCK converter with GPIO

Reviewer: ```suggestion
This sample uses the npmx BUCK driver to do the following:
  * Set output voltage for the BUCK converter. 
  * Set retention voltage for the BUCK converter with GPIO.
  * Control enable mode for the BUCK converter with GPIO.
```

> Voltage can be changed only for BUCK1. BUCK2 voltage is set by default to 3.0 V.

Reviewer: ```suggestion
Voltage can be changed only for BUCK 1. BUCK 2 voltage is set by default to 3.0 V.
```

Reviewer 3: And remember, each sentence on a separate line.

  .. list-table:: nPM1300 EK connections.
    :widths: 25 25 25
    :header-rows: 1

    * - nPM1300 EK pins
      - nRF5340 DK pins
      - nRF52840 DK pins
    * - GPIO0
      - P1.10
      - P1.10
    * - SDA
      - P1.02
      - P0.26
    * - SCL
      - P1.03
      - P0.27
    * - VOUT2 & GND
      - External supply (P21)
      - External supply (P21)

#. Make the following connections on the nPM1300 EK:

  * Connect a USB power supply to the **J3** connector.
  * Connect a suitable battery to the **J2** connector.
  * On the **P18** pin header, connect **VOUT2** and **VDDIO** pins with a jumper.
  * On the **P2** pin header, connect **VBAT** and **VBATIN** pins with a jumper.
  * On the **P17** pin header, connect all LEDs with jumpers.
  * On the **P13** pin header, connect **RSET1** and **GND** pins with a jumper.
  * On the **P14** pin header, connect **RSET2** and **VSET2** pins with a jumper.

#. Make the following connections on the chosen DK:

  * Set the **SW9** nRF power source switch to **VDD**.
  * Set the **SW10** VEXT → nRF switch to **ON**.

Configuration
*************

|config|

Configuration options
=====================

> Check and configure the following Kconfig options to specify which testcase you want to run:

Reviewer 3: You are right about that Marcin. Our template needs refreshing.

.. CONFIG_TESTCASE_SET_BUCK_VOLTAGE:

CONFIG_TESTCASE_SET_BUCK_VOLTAGE
  This sample configuration enables a simple output voltage testcase.

>   The voltage is set to 2.4 V (`NPMX_BUCK_VOLTAGE_2V4`) on BUCK 1.

Author: It's an enum in npmx_buck.h so part of npmx. Added tag.

  The output voltage of BUCK 1 is set to 3.3 V (:c:enumerator:`NPMX_BUCK_VOLTAGE_3V3`).
  When you change the input state of the selected GPIO, the voltage in the BUCK 1 output will change from 0 V to the previously selected voltage.

.. note::

  Only one Kconfig option may be enabled at a time.

Building and running
********************

.. |sample path| replace:: :file:`samples/pmic/npmx/buck`

.. include:: /includes/build_and_run.txt

Testing
=======

After programming the sample to your development kit, test it by verifying the voltage.

#. Connect a multimeter to VOUT1 that is the output voltage pin of BUCK1.
#. |connect_kit|
#. |connect_terminal|
#. Depending on the chosen testcase, do the following:

  * For simple output voltage testcase: Check output voltage on EK VOUT1 - it should be 2.4 V.
  * For advanced output voltage testcase: Check output voltage on EK VOUT1 - it should start at 1.0 V and increase every 200 ms up to 3.3 V.
  * For retention voltage testcase: Connect EK GPIO1 to GND. Check output voltage on EK VOUT1 - it should be 1.7 V. Connect EK GPIO1 to VDD. Check output voltage on EK VOUT1 - it should be 3.3 V.

>   * For enable bucks via pin testcase: Connect EK GPIO3 to GND. Check output voltage on EK VOUT1 - it should be 0.0 V. Connect EK GPIO3 to VDD. Check output voltage on EK VOUT1 - it should be 3.3 V.

Reviewer: Please also change 'via' to 'using'.

.. note::

   Only one Kconfig option may be enabled at a time.

Building and running
********************

.. |sample path| replace:: :file:`samples/pmic/npmx/buck`

.. include:: /includes/build_and_run.txt

Testing
=======

After programming the sample to your development kit, test it by verifying the voltage.

#. Connect a multimeter to VOUT1 that is the output voltage pin of BUCK 1.
#. |connect_kit|
#. |connect_terminal|
#. Depending on the chosen testcase, do the following:

   * For simple output voltage testcase: Check output voltage on EK **VOUT1** pin - it should be 2.4 V.
   * For advanced output voltage testcase: Check output voltage on EK **VOUT1** pin - it should start from 1.0 V and increase by 200 ms up to 3.3 V.
   * For retention voltage testcase: Connect EK **GPIO1** pin to GND.
     Check output voltage on EK **VOUT1** pin - it should be 1.7 V.
     Connect EK **GPIO1** pin to VDD.
     Check output voltage on EK **VOUT1** pin - it should be 3.3 V.
   * For enable bucks using pin testcase: Connect EK **GPIO3** pin to GND.

>      Check output voltage on EK **VOUT1** pin - it should be 0.0 V. Connect EK **GPIO3** pin to VDD.

Reviewer: ```suggestion
     Check output voltage on EK **VOUT1** pin - it should be 0.0 V.
     Connect EK **GPIO3** pin to VDD.
```

.. _npmx_buck_sample:

npmx: BUCK
##########

.. contents::
   :local:
   :depth: 2

The npmx BUCK sample demonstrates how to control PMIC BUCKs using npmx drivers.

Requirements
************

The sample supports the following development kits:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf52840dk_nrf52840, nrf5340dk_nrf5340_cpuapp

The sample also requires an nPM1300 EK.

Overview
********

This sample uses the npmx BUCK driver to do the following:

>   * Set output voltage for the BUCK converter.
>   * Set retention voltage for the BUCK converter with GPIO.
>   * Control enable mode for the BUCK converter with GPIO.

Reviewer: ```suggestion
* Set output voltage for the BUCK converter.
* Set retention voltage for the BUCK converter with GPIO.
* Control enable mode for the BUCK converter with GPIO.
```
